autogluon/utils/tabular/features/abstract_feature_generator.py

```diff
@@ -441,5 +295,128 @@ def drop_duplicate_features(X):
+
+                if (len(interval_index) >= min_desired_bins) and (len(interval_index) <= max_desired_bins):
+                    is_satisfied = True
+                    # print('satisfied', column, len(interval_index))
+                cur_iteration += 1
+                if cur_iteration >= max_iterations:
+                    is_satisfied = True
+                    # print('max_iterations met, stopping prior to satisfaction!', column, len(interval_index))
+
+            bin_mapping[column] = interval_index
+        return bin_mapping
+
+    # TODO: Move this outside of here
+    # TODO: Not accurate for categoricals, will count categorical mapping dict as taking more memory than it actually does.
+    @staticmethod
+    def get_approximate_df_mem_usage(df, sample_ratio=0.2):
+        if sample_ratio >= 1:
+            return df.memory_usage(deep=True)
+        else:
+            num_rows = len(df)
+            num_rows_sample = math.ceil(sample_ratio * num_rows)
+            return df.head(num_rows_sample).memory_usage(deep=True) / sample_ratio
+
+    # TODO: Clean code
+    # bins is a sorted int/float series, ascending=True
+    @staticmethod
+    def get_bins(bins: Series, bin_index, bin_epsilon):
+        max_val = bins.max()
+        bins_2 = bins.iloc[bin_index]
+        bins_3 = list(bins_2.values)
+        bins_unique = sorted(list(set(bins_3)))
+        bins_with_epsilon_max = set([i for i in bins_unique] + [i - bin_epsilon for i in bins_unique if i == max_val])
+        removal_bins = set([bins_unique[index - 1] for index, i in enumerate(bins_unique[1:], start=1) if i == max_val])
+        bins_4 = sorted(list(bins_with_epsilon_max - removal_bins))
+        bins_5 = [np.inf if (x == max_val) else x for x in bins_4]
+        bins_6 = sorted(list(set([-np.inf] + bins_5 + [np.inf])))
+        bins_7 = [(bins_6[i], bins_6[i + 1]) for i in range(len(bins_6) - 1)]
+        interval_index = pd.IntervalIndex.from_tuples(bins_7)
+        return interval_index
+
+    def print_feature_type_info(self):
+        logger.log(20, 'Original Features (raw dtypes):')
+        for key, val in self.feature_type_family_init_raw.items():
+            if val: logger.log(20, '\t%s features: %s' % (key, len(val)))
+        logger.log(20, 'Original Features (inferred dtypes):')
+        for key, val in self.feature_type_family.items():
+            if val: logger.log(20, '\t%s features: %s' % (key, len(val)))
+        logger.log(20, 'Generated Features (special dtypes):')
+        for key, val in self.feature_types_metadata.feature_types_special.items():
+            if val: logger.log(20, '\t%s features: %s' % (key, len(val)))
+        logger.log(20, 'Final Features (raw dtypes):')
+        for key, val in self.feature_types_metadata.feature_types_raw.items():
+            if val: logger.log(20, '\t%s features: %s' % (key, len(val)))
+        logger.log(20, 'Final Features:')
```

Comment:
I think all 4 print statements ("... Features ... :") should make it more clear what the difference between these 4 feature groups is.
Eg. user may not understand what's the difference between "Final Features (raw dtypes)" and "Final Features" or the difference between "raw dtypes" and "inferred dtypes"

Reply:
I gave it a shot but couldn't figure out a nice way to communicate it without making the logs overly verbose. Feel free to adjust these logs in a separate PR if you have suggestions though.